### DISCUSSION TITLE
Npm publish - secrets.GITHUB_TOKEN not used because .npmrc exists

### DISCUSSION THREAD

**Discussion by elexisvenator** [community/community#25414] [Packages]
<p>I have the following workflow to create and publish a npm package to github packages. It was copied verbatim from another repository where it works without issue.  However for this repo it says that the token (the built in <code>GITHUB_TOKEN</code>) is missing scopes.</p>
<pre><code class="lang-yaml"># This workflow will run tests using node and then publish a package to GitHub Packages when a release is created
# For more information see: https://help.github.com/actions/language-and-framework-guides/publishing-nodejs-packages

name: Publish

on:
  push:
    branches:
    - master

jobs:
  release:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v1
        with:
          node-version: 12
          registry-url: https://npm.pkg.github.com/
          scope: '@pageuppeopleorg'
      - run: npm i
      - run: npm run create
      - name: Bump version and push tag
        id: tag
        uses: anothrNick/github-tag-action@1.17.2
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          WITH_V: true
          RELEASE_BRANCHES: master
          DEFAULT_BUMP: patch
      - name: version
        run: npm --no-git-tag-version --allow-same-version version ${{ steps.tag.outputs.tag }}
      - name: publish
        run: npm publish
        env:
          NODE_AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}

</code></pre>
<p>Result of the <code>publish</code> step:</p>
<pre><code class="lang-auto">npm notice 
npm ERR! code E401
npm ERR! 401 Unauthorized - PUT https://npm.pkg.github.com/@pageuppeopleorg%2fpageup-ui-gateway-deploy-tool - Your token has not been granted the required scopes to execute this query. The 'createPackageVersion' field requires one of the following scopes: ['write:packages'], but your token has only been granted the: ['read:packages', 'repo'] scopes. Please modify your token's scopes at: https://github.com/settings/tokens.

</code></pre>
<p>As far as I am aware, the <code>secrets.GITHUB_TOKEN</code> should have <code>packages:write</code>. So I have no idea what i going on here.  Anyone seen something similar?</p>

**Comment by elexisvenator** [comment]
<p><strong>UPDATE:</strong> Turns out the issue is that i had a .npmrc file checked in. The file has a token that granted read package access to github packages.</p>
<p>Checking this file in was intentional and a pattern we use in our private repositories. So I will change the question to: How do I get npm to use the <code>secrets.GITHUB_TOKEN</code> when a .npmrc file already exists?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/elexisvenator">@elexisvenator</a>,</p>
<p>At first, you need to understand the restrictions of the <strong>GITHUB_TOKEN</strong>:</p>
<blockquote>
<p>The permissions of <strong>GITHUB_TOKEN</strong> are limited to the repository that contains your workflow.<br>
If you need a token that requires permissions that aren’t available in the <strong>GITHUB_TOKEN</strong>, you can create a personal access token and set it as a secret in your repository.</p>
</blockquote>
<p>This means the scope of the <strong>GITHUB_TOKEN</strong> is only inside of the repository where the current workflow is running. If the scope you require is beyond the current repository, you need to create a personal access token to obtain more scopes.<br>
You can reference the docs about <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token" rel="nofollow noopener"><strong>Authenticating with the GITHUB_TOKEN</strong></a> to view more details.</p>
<p>If you have configured a .npmrc file for your npm package and want to use the <strong>GITHUB_TOKEN</strong> in the <strong>.npmrc</strong> file, you need to do the steps below:</p>
<ol>
<li>In the workflow file,</li>
</ol>
<ul>
<li>when setup node, do not set <strong>registry-url</strong> on the <strong>setup-node</strong> action.</li>
<li>set <strong>secrets.GITHUB_TOKEN</strong> as an environment variable in the workflow.</li>
</ul>
<pre><code class="lang-yaml">- name: setup node
  uses: actions/setup-node@v1
  with:
    node-version: 12

. . .

- name: publish
  run: npm publish
  env:
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
</code></pre>
<ol start="2">
<li>In the .npmrc file use the variable name you set in the workflow file to access the environment variable for the <strong>GITHUB_TOKEN</strong>.</li>
</ol>
<pre><code class="lang-auto">//npm.pkg.github.com/:_authToken=${GITHUB_TOKEN}
</code></pre>
<p>The following is a simple example as reference.:<br>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.npmrc" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.npmrc" target="_blank" rel="nofollow noopener">TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.npmrc</a></h4>
<pre><code class="lang-npmrc">//npm.pkg.github.com/:_authToken=${GITHUB_TOKEN}
registry=https://npm.pkg.github.com/testgithubpackagesrml
</code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<br>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.github/workflows/CI.yml" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.github/workflows/CI.yml" target="_blank" rel="nofollow noopener">TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.github/workflows/CI.yml</a></h4>
<pre><code class="lang-yml"># This workflow will run tests using node and then publish a package to GitHub Packages when a release is created
# For more information see: https://help.github.com/actions/language-and-framework-guides/publishing-nodejs-packages

name: CI

on: push

env:
  MAIN_VER: 1

jobs:
  publish_gpr:
    name: Publish to GitHub Packages Registry
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v2

      - name: Setup node
        uses: actions/setup-node@v1
</code></pre>

  This file has been truncated. <a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.github/workflows/CI.yml" target="_blank" rel="nofollow noopener">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
</p>

**Comment by elexisvenator** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a>,<br>
Thanks for the reply :slight_smile: The .npmrc file has a different token in it, more restrictive than the <strong>GITHUB_TOKEN</strong>. I want to leave that there, but for the the github action workflow (specifically the publish step) I want to override the usage of that token with the GITHUB_TOKEN instead.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/elexisvenator">@elexisvenator</a>,</p>
<p>You can try to use the <strong>sed</strong> command to replace the token with the <strong>GITHUB_TOKEN</strong> in the <strong>.npmrc</strong> file, before executing the publish step. Here you also may need using <strong>regular expression</strong> in the <strong>sed</strong> command.<br>
A simple example as reference:</p>
<pre><code class="lang-yaml">- name: replace token
  run: echo "$(sed 's/\(_authToken=\)[a-z0-9]*\($\)/\1${{ env.GITHUB_TOKEN }}\2/g' .npmrc)" &gt; .npmrc
  env:
    GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
</code></pre>

**Comment by elexisvenator** [comment]
<p>Thanks <a class="mention" href="/brightran">@brightran</a>, the sed command worked for me :smiley:</p>

**Comment by vprasanth** [comment]
Using just wrap a `npm preinstall` step to `echo //npm.pkg.github.com/:_authToken=$GITHUB_TOKEN >> .npmrc `
